rfcs/0000-simple_unwind_annotation.md

```diff
@@ -1,179 +1,311 @@
+
+## Unwinding from a Rust callback in C: "C unwind" calls Rust
+
+Some C libraries like `mozjpeg` take callbacks using the platform native ABI and
+expect these callbacks to unwind through C back into the user application to
+report errors.
+
+JIT compilers like [Lucet][lucet] and [Weld][weld]: TODO (I'm not sure I understood what these programs do; need some help here)
+  * they generate code that calls other code using the C ABI
+  * all the code except the ABI is implemented in Rust, so they want to be able to unwind through the ABI
+  * the `"Rust"` call ABI is too "unstable" for their uses, and they prefer to use the more stable platform ABI
+
+[lucet]: https://github.com/fastly/lucet	
 [weld]: https://www.weld.rs/
 
-## Less dependence on C/C++
-
-Some of the [proposed workarounds][cffi-panic] to the lack of cross-FFI unwinding require the use of
-wrappers written in C (for `setjmp` and `longjmp`) or C++ (for exceptions). This solution reduces
-the amount of non-Rust code that must be generated and maintained when the foreign code is
-compatible with the unspecified Rust unwinding mechanism.
-
-[cffi-panic]: https://github.com/gnzlbg/cffi-panic
-
 # Guide-level explanation
 [guide-level-explanation]: #guide-level-explanation
 
-Rust function definitions with the `"C panic"` ABI string (e.g., `extern "C panic" fn`) are
-permitted to unwind with a panic, as opposed to `extern "C" fn` functions which will abort the
-process if a panic reaches the function boundary.
-
-When used on declarations of imported functions (e.g., `extern "C panic" { fn ... }`) or function
-pointers (e.g., `extern "C panic" fn()`), the `"C panic"` ABI string means that if the function
-unwinds, the unwind will be propagated though any calling code. If an `extern "C"` imported function
-or function pointer unwinds, the behavior is undefined.
+This RFC introduces a new ABI string called `"C unwind"` on some specific
+targets. If this ABI is not available for a target, Rust will error at
+compile-time saying so. This RFC adds support of this ABI for most targets with
+unwinding support, but it is expected that support for this ABI will be added to
+more targets in the future, either in other RFCs that build on this one, or in
+FCPs.
+
+Like for all other ABI strings, one can use this ABI string:
+
+  * to define a function in Rust: `extern "C unwind" fn foo() { }`
+  * to declare a function in Rust: `extern "C unwind" { fn foo(); }`
+  * in function pointer types: `let _: extern "C unwind" fn = foo;`
+    * Since `"C unwind"` is a super-set of `"C"`, there are implicit coercions
+      from `extern "C"` function types to `extern "C unwind"` function types.
+      These are analogous to the coercion rules for safe `fn` types to `unsafe
+      fn` types.
+
+Similarly to the `"Rust"` ABI string, `"C unwind"` functions can unwind:
+
+* if the `"C unwind"` function is defined in Rust, it unwinds the stack, and the
+  panic can be caught with `catch_unwind`.
```

Comment:
I presume you mean *and the panic originated in Rust* or something like that?

Reply:
Indeed, thanks.